### ISSUE TITLE
Restlet Resources in Blueprint (Apache Karaf)

### ISSUE THREAD

**Issue by kugtong33** [restlet/restlet-framework-java#1171]
I am using `Apache Karaf` and `Blueprint` , I am having a hard time integrating these technologies with Restlet. I found the `org.restlet.ext.osgi` library but not that many on how to use Providers using Blueprint.

So a sample `ServerResource` would be

``` java
public SampleResource extends ServerResource {
    @Get
    public Representation represent(Representation representation){
         /* Do something crazy here */
    }
}
```

to register it to a resource I must attach it to a router or an application

``` java
public class SampleRouter extends Router {
    public SampleRouter {
        super();
        attach("/sampleroute", SampleResource.class);
    }
}
```

attaching it using `SampleResource.class` is quite troublesome if I have dependencies injected using `Blueprint`. How can I create a `ServerResource` even if I have injected `Blueprint Services` or are there any alternatives?

**Comment by ankurnema** [comment]
Hello,

I also tried integrating Restlet with osgi but was not able to do so. Due
to deadline as of now I have integrated Apache Karaf and Restlet using
Component and SE approach . If you are interested I can share my approach.

On Mon, Nov 23, 2015 at 2:48 PM, Arjay Sitoy <notifications@github.com>
wrote:

> I am using Apache Karaf and Blueprint , I am having a hard time
> integrating these technologies with Restlet. I found the
> org.restlet.ext.osgi library but not that many on how to use Providers
> using Blueprint.
>
> So a sample ServerResource would be
>
> public SampleResource extends ServerResource {
>     @Get
>     public Representation represent(Representation representation){
>          /* Do something crazy here */
>     }
> }
>
> to register it to a resource I must attach it to a router or an application
>
> public class SampleRouter extends Router {
>     public SampleRouter {
>         super();
>         attach("/sampleroute", SampleResource.class);
>     }
> }
>
> attaching it using SampleResource.class is quite troublesome if I have
> dependencies injected using Blueprint. How can I create a ServerResource
> even if I have injected Blueprint Services or are there any alternatives?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/restlet/restlet-framework-java/issues/1171>.
>



-- 

Ankur Nema | Mobile: +91 8446221212 | ankurnema@gmail.com

**Comment by kugtong33** [comment]
@ankurnema , I would love to know your ideas regarding in this issue.

**Comment by ankurnema** [comment]
This is how I have designed my Osgi Bundle:

* I have created Configuration for getting port which I want to use for Restlet.
* I have created a Restlet Component which will attach a default application to default host. This default application in my case does nothing but send out a message "Service Not available".
* My OSGi Bundle will expose OSGi service which can be used to register different applications.
```Java
public interface MyCoreWSService
{
	public void attachApplicationToDefaultHost(String appName, EsmfWSApplication app);
}
```
* You can create component using following code

```Java
		wsComponent = new MyCoreWebserviceComponent();
		wsComponent.getServers().add(Protocol.HTTP, port);
		wsComponent.start();
```
* One can also use AJP in place of HTTP.
* I am maintaining map of application which other bundles register via MyCoreWSService for tracking all the applications and once wsComponent is started I attach existing applications to Default Host.

* If at runtime port is changed then I first Stop Component. Then wait for it to stop properly and then change port and re attach existing applications again.

* I have used apache as web server and used mod_proxy to move my calls to webservice backend. so if a call comes to http://localhost/ws it is proxied to http://localhost:8081/ws.

**Comment by kugtong33** [comment]
@ankurnema, I did some experiments on `blueprint` and it is somewhat the same on what you did, hope you find a solution on attaching `ServerResource` classes with injected dependencies also.

I came across with this tutorial [OSGi at REST][1] by @BryanHunt , it uses `Service Component Runtime (SCR)` and `org.restlet.ext.osgi`. I hope someone can confirm if it is also advisable for `blueprint`.

[1]:https://bryanhunt.wordpress.com/2011/06/30/osgi-at-rest/
